### PR TITLE
[Logs][Fix] Simplify logs tagging logic + add waiting for k8s tags

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

- Remove the local tag cache in the logs agent, rely on the tagger caching instead
- Resolve missing kubelet tags on initial logs caused by a race between AD and Tagger by introducing a warmup duration (disabled by default)

### Motivation

- Simplify the logs Tag Provider
- Ensure that first logs of a container are tagged with all kubernetes tags

### Additional Notes

- The logs agent calls the Tagger for each logs message now (similar to Dogstatsd)
- The tagger warmup duration fix is temporary, a planned PodWatcher revamp would resolve the race by design
- The tagger warmup duration wait is disabled by default

### CODE REVIEW

File: pkg/logs/tag/provider.go
Comment (xornivore) [inline pkg/logs/tag/provider.go]: Don't think you need these - it's the default value.

**Review by xornivore — DISMISSED** [review]

File: pkg/logs/tag/provider.go
Comment (olivielpeau) [inline pkg/logs/tag/provider.go]: these 2 lines will change the function from being entirely non-blocking to being blocking for 1 second + waiting until `updateTags` completes. Is that OK for the caller, or could this have undesirable effects upstream?

File: pkg/logs/tag/provider.go
Comment (olivielpeau) [inline pkg/logs/tag/provider.go]: how costly is a call to `updateTags`? If it's a little costly and we're handling a lot of log entries in the first few seconds, calling this for every new log entry could quickly become very costly.

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: the side effect is having the first few logs a bit late - but correctly tagged, our solutions engineers will communicate that to customers when suggesting this fix.

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: `updateTags` calls the tagger and the tagger itself uses the `cache` pkg to cache pods

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: in case of cache miss, the tagger queries the kubelet (users can enable a retry duration until the tagger find the container in the kubelet) but it's zero by default. once the entity is found in the tagger all the next tagger calls for log entries will hit the cache

Comment (olivielpeau) [reply pkg/logs/tag/provider.go]: ok, thanks for the explanation. I'm surprised there's a cache here if the tagger has its own cache (and that cache is relatively efficient already). Might be interesting to dig into why this cache was added in the first place

File: pkg/logs/tag/provider.go
Comment (olivielpeau) [reply pkg/logs/tag/provider.go]: Have you checked whether this affects other tailers or sources as well, on the caller? I'm not familiar with the logs codebase but in general I'd be cautious around adding delays in methods that the caller assumes are non-blocking.

In the same vein, to keep `Start` non-blocking unless `logs_config.k8s_wait_for_tags` is enabled, can we keep `updateTags` in the goroutine started below?

File: pkg/logs/tag/provider.go
Comment (olivielpeau) [inline pkg/logs/tag/provider.go]: I believe `pkg/logs` only accesses the config directly from `pkg/logs/agent.go` or `pkg/logs/config/`. Would make sense to keep this pattern here, but Logs-intake will have a more informed opinion on this.

**Review by olivielpeau — DISMISSED** [review]
A couple more questions. Approving for core, but would be interested in a review from logs-intake as well.

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: To be more precise the cache is used by the PodWatcher - called by the kubelet tag collectors. But yes I agree why this logs tag provider doesn't always query the tagger. Maybe @DataDog/logs-intake can answer this

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: sure thing, I'll ask @DataDog/logs-cold-storage

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: yes it affects other tailers but I do believe it's important to **ensure** we populate the local cache at least one time before calling the `GetTags` method (by calling `updateTags` before the `Start` method returns

Comment (olivielpeau) [reply pkg/logs/tag/provider.go]: hmm, to make sure I understand, are you concerned about the time it would take for the 1st call to `GetTags` to complete if the call it makes to `updateTags` has to populate the local cache?

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: My concern is when `k8s_wait_for_tags` is disabled, `GetTags` wouldn't be forced to call `updateTags` and returns the local cache directly. In this case there no guarantee the local cache has been populated.

Comment (olivielpeau) [reply pkg/logs/tag/provider.go]: ok I see, so to summarize:

* when `k8s_wait_for_tags` is disabled (default), we want to be absolutely sure the local cache is populated before `GetTags` is called, hence the blocking call to `updateTags` here.
* when `k8s_wait_for_tags` is enabled, we wait for 1 second (default) before this initial update, and then update on every call to `GetTags` for 5 seconds (default), to make sure we also get any new tags that the AD may find for this kubelet.

To be honest I don't what the side effects of making `Start` blocking are, but could you update the comment documenting the `Start` function, to reflect that it does a first blocking update of tags? And also update the PR description to reflect that this PR introduces this small change even with `k8s_wait_for_tags` left to its default value of `false` :)

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: That's correct, thanks for summarizing 👍 
I documented the `Start` method and updating the PR description to reflect that, and I also moved the wait for tagger duration when `k8s_wait_for_tags` is enabled) to the `GetTags` method instead of `Start`. This way we only block forwarding messages (in `tailer.forwardMessages()` instead of blocking the whole Start logic in the scheduler.

File: pkg/logs/tag/provider.go
Comment (ajacquemot) [reply pkg/logs/tag/provider.go]: this is correct, most of the times we inject dependencies for unit-tests and avoid accessing datadog config directly

File: pkg/logs/tag/provider.go
Comment (ajacquemot) [inline pkg/logs/tag/provider.go]: I think you could use once here: https://golang.org/pkg/sync/#Once

File: pkg/logs/tag/provider.go
Comment (ajacquemot) [inline pkg/logs/tag/provider.go]: Why do we need all those setters and getters ?

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: thanks for the trick

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: - `forceRefresh` is accessed by multiple functions in different goroutines, I implemented the getters and setters to ensure protecting it by the mutex

- `waitOnFirstCall` is no longer used and will be removed with its setter and getter (thanks to using `sync.Once`)

**Comment by ahmed-mez** [comment]
@olivielpeau @ajacquemot thanks for the comments. The PR is ready for a second review 🙇

File: pkg/logs/config/config.go
Comment (ajacquemot) [inline pkg/logs/config/config.go]: You could probably create a package variable for this

File: pkg/logs/tag/provider.go
Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: removed the local tags cache, as discussed with @ajacquemot

File: pkg/logs/config/config_test.go
Comment (ajacquemot) [inline pkg/logs/config/config_test.go]: let's both test the default value and an override please

File: releasenotes/notes/fix-logs-tagging-350289df776b181f.yaml
Comment (ajacquemot) [inline releasenotes/notes/fix-logs-tagging-350289df776b181f.yaml]:
```suggestion
    Configuring "logs_config.tagger_warmup_duration" delays the send of the first logs of a container.
```

File: releasenotes/notes/fix-logs-tagging-350289df776b181f.yaml
Comment (ajacquemot) [inline releasenotes/notes/fix-logs-tagging-350289df776b181f.yaml]: Not sure this section brings anything to the final users, I'd remove it.

**Review by ajacquemot — APPROVED** [review]
Looks pretty neats 👌, let's wait for an approval from someone from your team before merging please.

File: pkg/logs/tag/provider.go
Comment (mfpierre) [inline pkg/logs/tag/provider.go]: How would this behave with concurrency?

Let's say we call this twice in separate routines, the first call would get delayed but wouldn't the second one go through without the warmup being over, or am I missing something?

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: > Do calls the function f if and only if Do is being called for the first time for this instance of Once
(https://golang.org/pkg/sync/#Once) 

In our case, we create one config provider instance per AD service (entity).

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: So we guarantee that every first `GetTags` call for every Tag Provider will warmup

**Review by mfpierre — APPROVED** [review]

**Comment by ahmed-mez** [comment]
will merge once the CI pass

File: pkg/logs/tag/provider.go
Comment (mfpierre) [reply pkg/logs/tag/provider.go]: and for each AD service I guess the call is blocking other log lines from being processed right? (no concurrency inside a service)

Comment (ahmed-mez) [reply pkg/logs/tag/provider.go]: yes [forwardMessages()](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L116) runs in a single go routine (per service) and [doesn't](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L226-L254) create any other go routine 

Same for the [docker tailer](https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/docker/tailer.go)

@ajacquemot please correct me if I'm wrong

Comment (ajacquemot) [reply pkg/logs/tag/provider.go]: this is all correct, we could add a comment which states that this method is not thread safe and should not be called concurrently

**Pull request merged**
